packages/coreutils/src/plugins.ts

```diff
@@ -0,0 +1,867 @@
+   * the application is started.
+   *
+   * #### Notes
+   * The default is `false`.
+   */
+  autoStart?: boolean | 'defer';
+
+  /**
+   * The types of required services for the plugin, if any.
+   *
+   * #### Notes
+   * These tokens correspond to the services that are required by
+   * the plugin for correct operation.
+   *
+   * When the plugin is activated, a concrete instance of each type
+   * will be passed to the `activate()` function, in the order they
+   * are specified in the `requires` array.
+   */
+  requires?: Token<any>[];
+
+  /**
+   * The types of optional services for the plugin, if any.
+   *
+   * #### Notes
+   * These tokens correspond to the services that can be used by the
+   * plugin if available, but are not necessarily required.
+   *
+   * The optional services will be passed to the `activate()` function
+   * following all required services. If an optional service cannot be
+   * resolved, `null` will be passed in its place.
+   */
+  optional?: Token<any>[];
+
+  /**
+   * The type of service provided by the plugin, if any.
+   *
+   * #### Notes
+   * This token corresponds to the service exported by the plugin.
+   *
+   * When the plugin is activated, the return value of `activate()`
+   * is used as the concrete instance of the type.
+   */
+  provides?: Token<U> | null;
+
+  /**
+   * A function invoked to activate the plugin.
+   *
+   * @param app - The application provided by {@link PluginRegistry.application} .
+   *
+   * @param args - The services specified by the `requires` property.
+   *
+   * @returns The provided service, or a promise to the service.
+   *
+   * #### Notes
+   * This function will be called whenever the plugin is manually
+   * activated, or when another plugin being activated requires
+   * the service it provides.
+   *
+   * This function will not be called unless all of its required
+   * services can be fulfilled.
+   */
+  activate: (app: T, ...args: any[]) => U | Promise<U>;
+
+  /**
+   * A function invoked to deactivate the plugin.
+   *
+   * @param app - The application {@link PluginRegistry.application} .
+   *
+   * @param args - The services specified by the `requires` property.
+   */
+  deactivate?: ((app: T, ...args: any[]) => void | Promise<void>) | null;
+}
+
+/**
+ * Plugin registry.
+ */
+export class PluginRegistry<T = any> {
+  constructor(options: PluginRegistry.IOptions = {}) {
+    if (options.validatePlugin) {
+      console.info(
+        'Plugins may be rejected by the custom validation plugin method.'
+      );
+      this._validatePlugin = options.validatePlugin;
+    }
+  }
+
+  /**
+   * The application object.
+   *
+   * It will be provided as first argument to the
+   * plugins activation and deactivation functions.
+   *
+   * It can only be set once.
+   *
+   * By default, it is `null`.
+   */
+  get application(): T {
+    return this._application;
+  }
+  set application(v: any) {
```

Comment:
Shouldn't we type the parameter here ?
```suggestion
  set application(v: T) {
```

Reply:
Indeed it makes sense